app-example/src/main/java/com/badoo/ribs/example/photo_details/feature/PhotoDetailsFeature.kt

```diff
@@ -0,0 +1,126 @@
+        object LikeSent : Effect()
+        object UnlikeSent : Effect()
+        object LikeNotSent : Effect()
+    }
+
+    sealed class News {
+        object ShowLogin : News()
+    }
+
+    class BootStrapperImpl : Bootstrapper<Wish> {
+        override fun invoke(): Observable<Wish> = Wish.LoadPhoto.toObservable()
+    }
+
+    class ActorImpl(
+        private val photoDetailsDataSource: PhotoDetailsDataSource,
+        private val photoId: String,
+        private val authDataSource: AuthDataSource
+    ) : Actor<State, Wish, Effect> {
+        override fun invoke(state: State, wish: Wish): Observable<Effect> =
+            when (wish) {
+                Wish.LoadPhoto -> photoDetailsDataSource.getPhoto(photoId)
+                    .map<Effect> { Effect.PhotoLoaded(it) }
+                    .toObservable()
+                    .startWith(Effect.LoadingStarted)
+                Wish.LikePhoto -> likeOrDislikePhoto(state)
+            }
+
+        private fun likeOrDislikePhoto(state: State): Observable<Effect> =
+            when {
+                state !is State.Loaded -> Observable.empty()
+                authDataSource.getState() !is Authenticated -> Effect.LikeNotSent.toObservable()
+                else -> {
+                    if (state.detail.likedByUser) {
+                        unlikePhoto()
+                    } else {
+                        likePhoto()
+                    }
+                }
+            }
+
+        private fun likePhoto(): Observable<Effect> {
+            return photoDetailsDataSource.likePhoto(photoId)
+                .toObservable<Effect>()
+                .startWith(Effect.LikeSent)
+        }
+
+        private fun unlikePhoto(): Observable<Effect> {
+            return photoDetailsDataSource.unlikePhoto(photoId)
+                .toObservable<Effect>()
+                .startWith(Effect.UnlikeSent)
+        }
+    }
+
+    class ReducerImpl : Reducer<State, Effect> {
+        override fun invoke(state: State, effect: Effect): State =
+            when (effect) {
+                is Effect.LoadingStarted -> State.Loading
+                is Effect.PhotoLoaded -> State.Loaded(effect.detail)
+                is Effect.LikeSent -> likeState(state, true)
+                is Effect.UnlikeSent -> likeState(state, false)
+                is Effect.LikeNotSent -> state
+            }
+
+        private fun likeState(state: State, isLiked: Boolean): State {
+            return if (state is State.Loaded) {
+                state.copy(detail = state.detail.copy(likedByUser = isLiked))
+            } else {
+                state
+            }
+        }
+    }
+
+    class NewsPublisherImpl(
+        private val authDataSource: AuthDataSource
+    ) : NewsPublisher<Wish, Effect, State, News> {
+        override fun invoke(wish: Wish, effect: Effect, state: State): News? =
+            when (effect) {
+                is Effect.LikeNotSent -> News.ShowLogin
```

Comment:
`LikeNotSent` as a name doesn't tell me why to show login? You know, maybe it was not sent for any other reason, like failed network, just based on the name. I suggest to express auth failure in the effect name.